### PR TITLE
Use aurora clone for contact rollups

### PR DESCRIPTION
**Pull request by sureshc**
This is the approach I'm considering for using a clone of the production database in place of the read-only replica to create the `contact_rollups_daily` table and populate it with batch aggregation queries.

There are a bunch of TODOs in here, and I haven't updated the Contact Rollups Validation component to query the database clone, yet, but wanted to get feedback on this and see if there are alternatives. I’ve manually tested each of the pieces of this solution manually in the Ruby console.  Looking for guidance on how to test holistically.

### CODE REVIEW

**Comment by wjordan** [comment]
> Looking for guidance on how to test holistically.

First, I still think it's tragic that we're even touching this code when its test coverage was ['temporarily' disabled](https://github.com/code-dot-org/code-dot-org/pull/13547) two years ago and has never been re-added. Emphatically, this should be the first step before making any (substantial) changes. This is serious technical debt that absolutely needs to be addressed.

Assuming that this work is out of scope due to pressing time constraints, my suggestion would be to try and only focus on testing the changed parts, isolating it from the rest of the existing code. That is, assume the existing code does what it should, and only test the 'new' parts in isolation to make sure the new part does what you expect it to. Given that the new code operates on RDS resources, this would probably involve using the AWS SDK's stubbing functionality.

**Comment by hacodeorg** [comment]
Adding to Will's point, in order to isolate, test and optimize this code (or further rearchitect it), I'll improve logging to time how long each step take. At least we will be more informed before prioritizing work.

**Comment by sureshc** [comment]
Policy Actions that our `daemonrole` needs to be granted in #28840:

- describe_db_clusters - arn:aws:rds:region:account-id:cluster-endpoint:_CDO.db_cluster_id_ / arn:aws:rds:region:account-id:cluster-endpoint:_CDO.db_cluster_id_-temporary-clone
- delete_db_instance
- delete_db_cluster
- restore_db_cluster_to_point_in_time
- describe_db_cluster_endpoint
- create_db_instance

Policy Actions that `daemonrole` needs, but can not be granted to specific Resources:
- describe_db_instances (called by db_instance_deleted and db_instance available waiters)

File: bin/cron/build_contact_rollups [hunk 1]
```diff
@@ -40,6 +49,90 @@ def main
...
+def wait_until_db_cluster_deleted(db_cluster_id, max_attempts, delay)
+  rds_client = Aws::RDS::Client.new
+  attempts = 0
+  while attempts <= max_attempts
+    # describe_db_cluster will Raise a DBClusterNotFound Error when the cluster has been deleted.
+    rds_client.describe_db_clusters({db_cluster_identifier: db_cluster_id})
+    attempts += 1
+    sleep delay
+  end
+  raise Error("Timeout after waiting #{max_attempts * delay} seconds for cluster #{DATABASE_CLUSTER_CLONE_ID} deletion to complete.")
```
Comment (uponthesun) [inline bin/cron/build_contact_rollups]: StandardError or RuntimeError? I don't think just Error exists in Ruby.

File: bin/cron/build_contact_rollups [hunk 2]
```diff
@@ -40,6 +49,90 @@ def main
...
+    end
+    ChatClient.message(CHAT_CHANNEL, "Deleting clone cluster - #{DATABASE_CLUSTER_CLONE_ID}")
+    rds_client.delete_db_cluster(
+      {
+        db_cluster_identifier: DATABASE_CLUSTER_CLONE_ID,
+        skip_final_snapshot: true,
+      }
+    )
+    wait_until_db_cluster_deleted(DATABASE_CLUSTER_CLONE_ID, 10, 60)
+  rescue Aws::RDS::Errors::DBClusterNotFoundFault => error
```
Comment (uponthesun) [inline bin/cron/build_contact_rollups]: Nitpick - should reduce the code in begin block to only the code this rescue block is intended for (the  `describe_db_clusters` call).

File: lib/cdo/contact_rollups.rb [hunk 3]
```diff
@@ -25,32 +26,15 @@ class ContactRollups
...
-    query_timeout: MAX_EXECUTION_TIME_SEC
-  )
-
-  # Connection to read from Dashboard reporting database.
-  DASHBOARD_REPORTING_DB_READER = sequel_connect(
-    CDO.dashboard_reporting_db_reader,
-    CDO.dashboard_reporting_db_reader,
-    query_timeout: MAX_EXECUTION_TIME_SEC
-  )
+  DATABASE_CLUSTER_CLONE_ID = "#{CDO.db_cluster_id}-temporary-clone"
```
Comment (uponthesun) [inline lib/cdo/contact_rollups.rb]: Can you share this variable with `build_contact_rollups` so it's not defined in two places?

File: lib/cdo/contact_rollups_validation.rb [hunk 4]
```diff
@@ -1,9 +1,9 @@
 require File.expand_path('../../../pegasus/src/env', __FILE__)
+require 'aws-sdk-rds'
 
-class ContactRollupsValidation
-  # Connection to read from Pegasus reporting database.
-  PEGASUS_REPORTING_DB_READER = sequel_connect(CDO.pegasus_reporting_db_reader, CDO.pegasus_reporting_db_reader)
+DATABASE_CLUSTER_CLONE_ID = "#{CDO.db_cluster_id}-temporary-clone"
```
Comment (uponthesun) [inline lib/cdo/contact_rollups_validation.rb]: And here as well.

File: lib/cdo/contact_rollups_validation.rb [hunk 5]
```diff
@@ -344,4 +345,26 @@ def self.log(s)
     # puts s unless Rails.env.test?
     CDO.log.info s
   end
+
+  private
+
+  def initialize_connection_to_database_clone
```
Comment (uponthesun) [inline lib/cdo/contact_rollups_validation.rb]: Reduce duplication with `initialize_connections_to_database_clone` in `contact_rollups.rb`?

**Review by uponthesun — APPROVED** [review]
LGTM, assuming you and Will are on the same page on testing (see his last comment).

File: bin/cron/build_contact_rollups (hunk 2 above)
Comment (sureshc) [reply bin/cron/build_contact_rollups]: I was intentionally using the Error to break out of the loop.  How does this change look?

File: lib/cdo/contact_rollups.rb (hunk 3 above)
Comment (sureshc) [reply lib/cdo/contact_rollups.rb]: Done!

File: lib/cdo/contact_rollups_validation.rb (hunk 5 above)
Comment (sureshc) [reply lib/cdo/contact_rollups_validation.rb]: Done!

**Pull request merged**